### ISSUE TITLE
Alternatives to Android Lifecycle issue 

### ISSUE THREAD

**Issue by sagrawal31** [razorpay/razorpay-cordova#121]
## Common to Both platforms

  - [x] I have gone through the README(https://github.com/razorpay/razorpay-cordova)
  - [x] I have searched for a similar issue (https://github.com/razorpay/razorpay-cordova/issues)
  - [x] I am using the latest version of our framework (https://github.com/razorpay/razorpay-cordova/releases)
  - [x] I have cleaned my project.
  - [x] I have gone through FAQ's (https://github.com/razorpay/razorpay-cordova/wiki/FAQ).

## Specific to iOS Users

*Not applicable on this issue as I need a discussion on alternatives for Android app*

## IDE Specs

*Not applicable on this issue as I need a discussion on alternatives for Android app*

## Retro Steps

*Not applicable on this issue as I need a discussion on alternatives for Android app*

## Problem Description

I have read the [Android Lifecycle Guide](https://github.com/razorpay/razorpay-cordova#android-lifecycle-guide) and tested the scenario of Android OS killing the original app.

I have reproduced this scenario by enabling `Don't keep activities` setting in the `Developer Options` on my Android device.

I know that this is how the Razorpay Android SDK is written, as I can see it [here](https://github.com/razorpay/razorpay-cordova/blob/v1.4.6/src/android/Main.java#L33) and we can't control the behaviour of Android OS or rather can't prevent it.

Since our primary users of the app are going to be college students so the probability of our app being used in low-end devices is very high that means as soon as a student goes for the payment screen, the app will get killed in the background and once they complete the payment (or closes the modal), they will see the app being relaunched.

So theoretically, we can't use this Cordova plugin directly in our Cordova app. Neither our client's business allowing this behaviour to happen. Yes, we can manage the state and do things around but it still breaks the behaviour and user experience.

## Solution or alternative we took

As an alternative, I removed this plugin and I modified the proxy code (which is used in the browser platform) https://github.com/razorpay/razorpay-cordova/blob/v1.4.6/src/browser/CheckoutProxy.js and wrapped it for my Angular/Ionic application.

Here is the link so that someone else can use https://gist.github.com/sagrawal31/3471820b4f58024990cf888ae4c87377

**And it worked like charm when I ran it on the browser using `ionic serve`**. The payment response came back to me properly.

But when I see that Razorpay opens up a child window using `window.open`, I got a little worried if that will work or not on the mobile app and my fear came true.

As soon as I entered the card detail and went further, it tried to open a child window and the app crashed.

## Questions to discuss

1. Is there a way we can tell Razorpay checkout script to not open a child window and instead redirect in the same iframe? (I tried to see the minified code but couldn't find any such option).
2. Is this behaviour of opening a child window only there in the test mode or is it in the live mode too? If that is only in the test mode that we can survive probably.
3. Is there another normal checkout way where we can open a certain URL using [`cordova-plugin-inappbrowser`](https://github.com/apache/cordova-plugin-inappbrowser) and which also not open in a child window?

**I hope you guys will understand this blocker issue for our app launch because, here in India, we can't expect our college students to have a high-end Android device and we can't go live with the probability that the Android OS won't kill the app in the background.**

Sorry to say but if we do not get a proper alternative, we have to choose another payment gateway.

**Comment by sumedht** [comment]
@sagrawal31 Can u share crash logs of the issue?
also, have u faced this issue in real-time? we have added [this](https://github.com/razorpay/razorpay-cordova#android-lifecycle-guide) to handle lifecycle scenario issue.

**Comment by sagrawal31** [comment]
>  Can u share crash logs of the issue?

Can you please re-read my description. I never mentioned that "my app is crashing". 

> also, have u faced this issue in real-time?

Yes, on low-end devices, the original app activity is getting killed by Android OS which is not desirable for our business.

> we have added this to handle lifecycle scenario issue.

I know this lifecycle issue and I have read it multiple times. **THAT IS OUR PROBLEM**. We don't want our users to see that after payment, the app is being relaunched.

**Comment by vivekshindhe** [comment]
> Yes, on low-end devices, the original app activity is getting killed by Android OS which is not desirable for our business.

Can you share the list of devices and their information where you can see this happen in real-time?

**Comment by sagrawal31** [comment]
@vivekshindhe I don't really have to test it on a real device because I only enabled `Don't keep activities` in the `Developers Options` in the Android OS and the problem started appearing clearly.

(This I have already explained in my issue description).

And since there is a clear chance of this happening i.e. Android OS killing the main app/activity, we can't take that risk especially when our main users of the app are the students. (This I have also explained).

**Comment by sumedht** [comment]
@sagrawal31 
For this issue, we have to handle `event.pendingResult` which is supported in v5.1.0 and above.
We are currently on v3.4.0 and have to move to the latest. For that, there may be some impact on functionality.
As it's a hypothetical issue, we will take it up as a low priority and fix it. will update you once it is done. 
Thanks for pointing out on this!

**Comment by sumedht** [comment]
@sagrawal31 
For fixing this scenario, we have to handle `event.pendingResult` in `onResume` state which will give pending callbacks if there is any when activity restarted.
This feature is provided by the framework on v5.1.0+ and we are currently on v3.4.0.
We will update our SDK with latest framework to handle this scenario. But as it's a hypothetical issue we will keep its priority low and will update you once it is fixed.

Thanks for pointing out this scenario which may affect real-time transactions!

**Comment by sumedht** [comment]
@sagrawal31 Updated in our latest SDK version. Closing this issue as there is no further activity on this.

**Comment by sagrawal31** [comment]
Thanks @sumedht for the update.
